### ISSUE TITLE
What am I doing wrong? :P

### ISSUE THREAD

**Issue by pixeleet** [motorcyclejs/motorcyclejs#26]
Given the following files.
- index.js
- List.js
- Slider.js

Trying to reimplement: https://github.com/milankinen/culli/tree/master/perf
for motorcycle as to have a comparison with cycle itself, but I'm failing hard with `s.subscribe not being a function`. The strange part is that in the kitchen sink project I'm trying things out in the store and motorcycle just work well together, I'm wondering where this example might be wrong.

Please let me know if you had a minute to reproduce to issue or have a suggestion to fix it :) 

__index.js__
```
import {run} from '@motorcycle/run'
import {makeDomDriver} from '@motorcycle/dom'
import Store, {Memory} from '@culli/store'
import List, {nextId} from './List'
import { O } from '@culli/base'

const N = 10000
const sliders =
  Array.apply(0, Array(N)).map(() => ({ id: nextId(), val: Math.floor(10 + Math.random() * 80) }))

const domDriver = makeDomDriver(document.body)
const storeDriver = Store(Memory({items: sliders}))

function effects(sinks) {
  return {
    DOM: domDriver(sinks.DOM),
    Store: storeDriver(sinks.Store, O.Adapter)
  }
}

run(List, effects)
```

__List.js__
```
import * as O from 'most'
import Slider from './Slider'
import { h1, div, hr, button } from '@motorcycle/dom'
import { combineObj } from 'most-combineobj'
import isolate from '@cycle/isolate'

let ID = 0
export const nextId = () => ++ID

export default function main(sources) {
  const {Store} = sources
  const {dispatch, props} = model(Store)
  const {vdom, children} = view(props)
  const actions = intent(sources.DOM)

  return {
    DOM: vdom,
    Store: O.merge(dispatch(actions), children.Store)
  }

  function model({actions, value}) {
    const dispatch = actions.reduce((state, action) => {
      switch (action.type) {
        case 'ADD':
          return {items: [{id: nextId(), val: Math.floor(10 + Math.random() * 80)}, ...state.items]}
        default:
          return state
      }
    })

    return {
      dispatch,
      props: {
        items: value.select('items')
      }
    }
  }

  function view({items}) {
    const children = items.value.mapChildren(it => isolate(Slider)({...sources, Store: it}),
      {values: ['DOM'], events: ['Store']})
    const state = combineObj({ childrenDOM: children.DOM, items: items.value })
    const vdom = state.map(({ childrenDOM, items }) => div({}, [
      h1('.header', `CULLI witch cycle DOM: ${items.length}`),
      button('add', 'Add slider'),
      hr(),
      div('.sliders', childrenDOM)
    ]))

    return {vdom, children}
  }

  function intent(domSource) {
    return domSource
      .select('.add').events('click')
      .map(e => ({type: 'ADD'}))
  }
}
```

__Slider.js__
```
import { div, input } from '@motorcycle/dom'

export default function main({ DOM, Store, min = 0, max = 100, step = 1 }) {
  const {dispatch, props} = model(Store)
  const vdom = view(props, {min, max, step})
  const actions = intent(DOM)

  return {
    DOM: vdom,
    Store: dispatch(actions)
  }

  function model({actions, value}) {
    const dispatch = actions.reduce((state, action) => {
      switch (action.type) {
        case 'SET':
          return {...state, val: action.payload}
        default:
          return state
      }
    })

    return {
      dispatch,
      props: {
        val: value.select('val')
      }
    }
  }

  function view({val}, {min, max, step}) {
    return val.value.map(val => {
      return div({}, [
        input('.slider', { attrs: {type: 'range', min, max, step, value: val} }), ' ', val
      ])
    })
  }

  function intent(domSource) {
    return domSource
      .select('.slider').events('input')
      .map(e => ({type: 'SET', payload: e.target.value}))
  }
}
```

**Comment by TylorS** [comment]
I have no clue if Culli works with Motorcycle.js (pinging @milankinen) in it's current form. It's likely that it's making use of Cycle.js' use of adapters, which would likely cause some errors here given they are not present. 

Other than that it is really hard for me to pinpoint any errors here never having used Culli.

**Comment by pixeleet** [comment]
I was kinda hoping you'd see something inherently wrong with the code itself :)

**Comment by pixeleet** [comment]
I actually did get culli/store to work with motorcycle in the kitchen sink project I'm using to try stuff out, very curious as to where is that slight little difference which prevents this from working, will get back if I find out :)

**Comment by TylorS** [comment]
I don't see anything wrong with the code, but the example Culli has itself is written with Cycle.js + Most, so there shouldn't be much change necessary.

**Comment by milankinen** [comment]
Cheers!

Cool to see that people are interested in new stuff! But yeah, like @TylorS said there should not be any technical issues by using `@culli/store` with `motorcycle`, as long as you manually pass the stream adapter as a second arg to the driver function. 

The codes looks legit, there are only two places that I'd change:
* `@motorcycle/dom` doesn't seem to understand numbers as children, hence string conversion is needed (slider.js)
```diff
-input('.slider', { attrs: {type: 'range', min, max, step, value: val} }), ' ', val
+input('.slider', { attrs: {type: 'range', min, max, step, value: val} }), ' ', `${val}`
```

* Isolation scope is missing and may cause some problems or at least decreases app's performance (list.js)
```diff
-const children = items.value.mapChildren(it => isolate(Slider)({...sources, Store: it}),
+const children = items.value.mapChildren((it, id) => isolate(Slider, id)({...sources, Store: it}),
```

That said, I still didn't get your example running... 😕 Instead I get the following error when I try to update any slider (initial render works fine after the previous fixes)
<img width="676" alt="motorcycle_error" src="https://cloud.githubusercontent.com/assets/1080706/22713185/1197f6b8-ed90-11e6-8e8e-80aa10c77327.png">

I suspect there is something weird happening in Mostly DOM's hyperscript helper function, in [these lines](https://github.com/TylorS/mostly-dom/blob/2dcca4d97289415ce5f7664e0db15221b475287f/src/hyperscript/h.ts#L69-L70). Those derive the `vnode.scope` from parent (`div` element) to children (`text` node in this case) that doesn't have any `.setAttribute` function, causing the previous error during update (`updateElement` function tries to use `setAttribute` always if there is `.scope` present).

Do you have @TylorS any insight about this?

**Comment by pixeleet** [comment]
@milankinen I have no problems at all running motorcycle and the store together in a kitchen sink app I'm using to test the different implementations of cycle drivers and state management solutions, that's why I was suspecting that I'm doing something wrong in code above.

**Comment by TylorS** [comment]
Ahh thanks @milankinen for finding me a bug! Would you mind opening an issue? :smile:

**Comment by milankinen** [comment]
@pixeleet did you use `isolate` in your kitchen sink app?

@TylorS sure 😄

**Comment by pixeleet** [comment]
@milankinen indeed, I am, it was _just there_ :)

**Comment by milankinen** [comment]
Hmmm that's weird... Any change to share its code to see what's different compared to this case?

**Comment by pixeleet** [comment]
Let me create a small private repo to mess with :)

**Comment by milankinen** [comment]
@TylorS made a quick fix so I managed to [get it working](https://github.com/milankinen/culli/tree/master/perf/motorcycle-sum)! However, it seems very slow compared to "pure CULLI" and `onionify` ones... 😕

**Comment by TylorS** [comment]
I can't say why for instance, but I've moved away from my target focus (at least for now) on high performance to rethinking Motorcycle and moving towards a much more type-safe solution. For instance `run :: (Sources -> Sinks) -> (Sinks -> Sources)` instead of `run :: (Sources -> Sinks) -> { [ key ] : DriverFn }`. In many way's I'm no longer looking to be a sister project of Cycle.js, but move it in a direction I'm happier with myself, one can likely expect more changes in the future that move further from Cycle.js equivalents and on to a "newer" way of thinking. 

One thing for instance, I'm probably going to remove any mechanisms for `isolate` from Motorcycle solutions like `vNode.scope` from `mostly-dom` and `isolateSink` and `isolateSource` from `@motorcycle/dom`'s DomSource implementation, because I don't like the impure solution, and would much prefer explicit handling of those situations.

**Comment by milankinen** [comment]
>  I'm probably going to remove any mechanisms for isolate from Motorcycle solutions like vNode.scope from mostly-dom and isolateSink and isolateSource from @motorcycle/dom's DomSource implementation

👍 👍 Do you already have ideas how to achieve isolation without them? Or is isolation even needed anymore in your opinion?

**Comment by pixeleet** [comment]
@TylorS Can you expand on the "newer" thinking? :) Interested in what you have in mind

**Comment by TylorS** [comment]
@milankinen I simply don't feel it's Motorcycle's job to figure these things out for a user. I think the impure and opaque solution we have now also lends to (I've seen it in my dev teams) just wrapping `isolate()` around this and that and not actually consider *why* or *what* they are during that. I'm also become much wiser (I hope), if not only more opinionated, over the last year and really dislike impurity or mutation.

@pixeleet I'm exploring lots of different things locally, but I don't think there is much more improvement I can think of for `@motorcycle/run`, it's very much where I want it to be. Other tools like `@motorcycle/dom` and some other things I'm doing could likely be rethought. Each of the tools driver here `@motorcycle/dom` uses a library I wrote for rendering called `mostly-dom` instead of snabbdom, and `@motorcycle/history` uses a library I wrote called `prehistoric` instead of `history` because I care about code quality :wink:
